### PR TITLE
Issue 1841 pub flow - createAsset Fn

### LINKED ISSUE #1841: Add publish flow
Add `createAsset` helper function ( get inspiration [from tests](https://github.com/oceanprotocol/ocean.js/blob/c622b213ff19f091e6ce194c7dc512ee2cc40626/test/integration/helpers.ts#L21)) 
- include oasis integration, check if dt4, if so add FileObject in dt constructor
- add possibility to use template 1,2 or 4

### PR DESCRIPTION
**Pull request by paulo-ocean**
Fixes #1841 .

Changes proposed in this PR:

- generic function (exported) to create an asset
- update config, set confidential evm key
- calculate template index (from id or address)
- if confidential evm have filesObject on constructor (no need to provider encrypt)

### CODE REVIEW

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:155]: if filesObject is stored in the datatoken contract, there is no point of keeping it in ddo.  Especially unencrypted

**Review by alexcos20 — CHANGES_REQUESTED** [review]

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:99]: if network is confidential EVM.
We need to extend config with this extra key.

confidential EVM = Sapphire, Sapphire testnet.   In the future we will have more, like Ocean's L1 or OceanEnterprise chains

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:152]: if network is confidential EVM

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:89]: templateIndex should be calculated.

Don't go to "templateIndex == templateId", this will fail.

For instance, on sapphire we have the following config now:  (https://github.com/oceanprotocol/contracts/blob/main/addresses/address.json#L261C3-L282C5)

```
oasis_saphire_testnet": {
    "chainId": 23295,
    "Ocean": "0x973e69303259B0c2543a38665122b773D28405fB",
    "OPFCommunityFeeCollector": "0xe8c6Dc39602031A152440311e364818ba25C2Bc1",
    "startBlock": 2565487,
    "Router": "0x5FBC2A29f7C8e4533dECD14Dc1c561A8eF0d4e31",
    "FixedPrice": "0x0846aec7aADD547131484Bf7fE548589b7F057fe",
    "ERC20Template": {
      "1": "0x519d462294958A935D34351b3ad9F98cCAB0b7f8",
      "2": "0x2bCa501a6f9FB8a96f1b7B09D4f74A32ACAa7584",
      "3": "0xf00c16665b79F23c04e5c8652Fe9853046b987c8",
      "4": "0x89b752C3e272feBA17B363EaaBFe0d653cF75bE7",
      "5": "0xb4914dfeC7f70C708673620eDF6d8839F81E7612"
    },
    "ERC721Template": {
      "1": "0x9Ab15cc945b2B58A51384539211dE8a88a1355E8"
    },
    "Dispenser": "0xbde8224d15885c2E46Be665b9392E619cD142de8",
    "ERC721Factory": "0x315ef329D214cde5cE7C206104aC35B1fe66bb8c",
    "PredictoorHelper": "0x2e4cb19faD077768fDcB04Bc0549AB76BE8eA63F",
    "AccessListFactory": "0xdA412123a3Cc0D921bcf7246f8742e3b7c3E7b94"
  },
  ```
  
  Guess what?  templateIndex 4 is disabled..
  
  ```bash
  Using the following as start point:
{
  chainId: 23295,
  Ocean: '0x973e69303259B0c2543a38665122b773D28405fB',
  OPFCommunityFeeCollector: '0xe8c6Dc39602031A152440311e364818ba25C2Bc1',
  startBlock: 2565487,
  Router: '0x5FBC2A29f7C8e4533dECD14Dc1c561A8eF0d4e31',
  FixedPrice: '0x0846aec7aADD547131484Bf7fE548589b7F057fe',
  ERC20Template: {
    '1': '0x519d462294958A935D34351b3ad9F98cCAB0b7f8',
    '2': '0x2bCa501a6f9FB8a96f1b7B09D4f74A32ACAa7584',
    '3': '0xf00c16665b79F23c04e5c8652Fe9853046b987c8',
    '4': '0x89b752C3e272feBA17B363EaaBFe0d653cF75bE7',
    '5': '0xb4914dfeC7f70C708673620eDF6d8839F81E7612'
  },
  ERC721Template: { '1': '0x9Ab15cc945b2B58A51384539211dE8a88a1355E8' },
  Dispenser: '0xbde8224d15885c2E46Be665b9392E619cD142de8',
  ERC721Factory: '0x315ef329D214cde5cE7C206104aC35B1fe66bb8c',
  PredictoorHelper: '0x2e4cb19faD077768fDcB04Bc0549AB76BE8eA63F',
  AccessListFactory: '0xdA412123a3Cc0D921bcf7246f8742e3b7c3E7b94'
}
nftCount
BigNumber { value: "1" }
Router owner: 0xC7EC1970B09224B317c52d92f37F5e1E4fF6B687
currentTokenCount
BigNumber { value: "5" }


------------
templateIndex:1
[
  '0x519d462294958A935D34351b3ad9F98cCAB0b7f8',
  true,
  templateAddress: '0x519d462294958A935D34351b3ad9F98cCAB0b7f8',
  isActive: true
]
templateId: 1


------------
templateIndex:2
[
  '0x2bCa501a6f9FB8a96f1b7B09D4f74A32ACAa7584',
  true,
  templateAddress: '0x2bCa501a6f9FB8a96f1b7B09D4f74A32ACAa7584',
  isActive: true
]
templateId: 2


------------
templateIndex:3
[
  '0xf00c16665b79F23c04e5c8652Fe9853046b987c8',
  true,
  templateAddress: '0xf00c16665b79F23c04e5c8652Fe9853046b987c8',
  isActive: true
]
templateId: 3


------------
templateIndex:4
[
  '0x89b752C3e272feBA17B363EaaBFe0d653cF75bE7',
  false,
  templateAddress: '0x89b752C3e272feBA17B363EaaBFe0d653cF75bE7',
  isActive: false
]
templateId: 4


------------
templateIndex:5
[
  '0xb4914dfeC7f70C708673620eDF6d8839F81E7612',
  true,
  templateAddress: '0xb4914dfeC7f70C708673620eDF6d8839F81E7612',
  isActive: true
]
templateId: 4
```


so, the proper way of determining templateIndex is :

```js

    const currentTokenCount = await factoryERC721.connect(owner).getCurrentTemplateCount(options);
    for(i=1;i<=currentTokenCount;i++){
        const tokenTemplate = await factoryERC721.connect(owner).getTokenTemplate(i,options);
        console.log("\n\n------------\ntemplateIndex:"+i)
        console.log(tokenTemplate)
        const erc20Template = await ethers.getContractAt("ERC20Template", tokenTemplate["templateAddress"]);
        const id=await erc20Template.connect(owner).getId()
        console.log("templateId: "+id)
        if(tokenTemplate.isActive && id==SEARCHED_TEMPLATE_ID || tokenTemplate["templateAddress"]==SEARCHED_TEMPLATE_ADDRESS){
                               return i
        }
    }
    return null     
    ```

File: src/config/Config.ts
Comment (alexcos20) [inline src/config/Config.ts:50]: why do we need this?

File: src/config/ConfigHelper.ts
Comment (alexcos20) [inline src/config/ConfigHelper.ts:183]: why do we need this?
We should always compare or act based on chainId and not on network name

File: src/config/ConfigHelper.ts
Comment (alexcos20) [inline src/config/ConfigHelper.ts:205]: why do we need this? We can fetch the list from ERC721Factory

File: src/config/ConfigHelper.ts
Comment (alexcos20) [inline src/config/ConfigHelper.ts:221]: why do we need this? We can fetch the list from ERC721Factory

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:81]: why do we need this function?

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:181]: change to nftFactory.
but why? We can get nftFactory from addresses(env or contracts npm package), so we don't need it as param

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:298]:
```suggestion
else{
    ddo.services[0].files=null
}
```

Comment (alexcos20) [reply src/utils/Assets.ts:298]: if it's confidential, that we don't store the unencrypted URL in the ddo, cause anyone can grab it

File: test/integration/PublishFlows.test.ts
Comment (alexcos20) [inline test/integration/PublishFlows.test.ts:210]: I would a templateID here.
If you want to test with addresses, then read addresses.json and use one of the values.
But never hardcode it

File: src/utils/Assets.ts
Comment (alexcos20) [inline src/utils/Assets.ts:176]: rename to  templateIDorAddress pls, so everybody knows that they need to pass either an templateID (and not template Index), or templateContractAddress

File: test/unit/AssetUtils.test.ts
Comment (alexcos20) [inline test/unit/AssetUtils.test.ts:44]: Why do we need this?
Never assume any hardcoded address.
Maybe next week we will change the templates, cause we have found a bug.  Or deploy a new product..
Then all ocean.js tests will fail, although they shoudn't

File: test/unit/AssetUtils.test.ts
Comment (alexcos20) [inline test/unit/AssetUtils.test.ts:53]: see above

File: test/unit/AssetUtils.test.ts
Comment (alexcos20) [inline test/unit/AssetUtils.test.ts:59]: see above

File: test/unit/AssetUtils.test.ts
Comment (alexcos20) [inline test/unit/AssetUtils.test.ts:70]: see above

**Review by alexcos20 — CHANGES_REQUESTED** [review]

File: src/config/ConfigHelper.ts
Comment (paulo-ocean) [reply src/config/ConfigHelper.ts:183]: `getConfig`
`public getConfig(network: string | number, infuraProjectId?: string): Config {`
...
 uses either network name or chain id... If we supply and search for a "name"  we might not able to check if it is a confidential EVM... specially because we have wrong names on the contracts address.json  ('saphire' misspelled with only 1 'p'). Was trying to overcome this obstacle and make it more flexible, but of course i can put only chain ids on the constant

File: src/config/ConfigHelper.ts
Comment (paulo-ocean) [reply src/config/ConfigHelper.ts:205]: not needed, probably was thinking about something else at the time

**Comment by paulo-ocean** [comment]
btw, the service "files" cannot be null for confidential EVM, but empty string instead... 
if its null the DDO does not passes against schema validation
schema:ServiceShape
```
....
sh:targetClass schema:Service ;
    sh:property [
        sh:path schema:id ;
        sh:datatype xsd:string ;
        sh:pattern "^(.*)$" ;
        sh:minCount 1;
        sh:maxCount 1;
    ] ;
...
```

**Comment by alexcos20** [comment]
> btw, the service "files" cannot be null for confidential EVM, but empty string instead... if its null the DDO does not passes against schema validation schema:ServiceShape
> 
> ```
> ....
> sh:targetClass schema:Service ;
>     sh:property [
>         sh:path schema:id ;
>         sh:datatype xsd:string ;
>         sh:pattern "^(.*)$" ;
>         sh:minCount 1;
>         sh:maxCount 1;
>     ] ;
> ...
> ```

then let's go with empty string

File: test/integration/PublishFlows.test.ts
Comment (mariacarmina) [inline test/integration/PublishFlows.test.ts:220]: shall we use this code after merging this branch with main?

**Review by mariacarmina — COMMENTED** [review]
Hello @paulo-ocean can you also merge the main branch into this PR please?

File: test/integration/PublishFlows.test.ts
Comment (paulo-ocean) [reply test/integration/PublishFlows.test.ts]: not the console.log :-) I've cleaned the test. If you mean the new function createAsset, yes, its an option, or we can do it all in many steps as we we're doing before (like some of the other tests)

File: src/utils/Assets.ts
Comment (mariacarmina) [inline src/utils/Assets.ts:199]: **Observation**: If datatoken 4 has the file object and we want to create a bundled tx (meaning deploy NFT, datatoken 4 and pricing schema at once) and files object needs to be encrypted on chain, then we should also have a test when we check, if confidentialEVM flag is set on true we need to `sapphire.wrap()` the owner wallet address, otherwise the tx won't be encrypted on sapphire chain.
**Question**: When creating a datatoken from a standalone NFT, NFT class should also have this confidentialEVM flag when creating a datatoken template 4, right?

**Review by mariacarmina — COMMENTED** [review]
I have an observation and a question regarding datatoken 4 creation.

File: src/utils/Assets.ts
Comment (mariacarmina) [reply src/utils/Assets.ts:199]: for question, I have created https://github.com/oceanprotocol/ocean.js/pull/1859 in case.

Comment (paulo-ocean) [reply src/utils/Assets.ts:199]: `sapphire.wrap()` is a specific API for Oasis. At this point its true that its the only confidential EVM we have, but that could be different tomorrow, we **cannot**  do shapire.wrap() on a TX just because its "confidentialEVM". It only works for Oasis shappire. In any case i think that applies only for `calldata` when interacting with the SC, as the smart contract should run on their own secure enclave... not really sure if that's on the scope of this function here. Maybe @alexcos20  can have some input here.. its easy, but in any case we need to check the chain, its not enough see if its confidential or not.

Regarding your question, i'm not sure if i understand your point.. When we have a confidential EVM, the requirement would be pass the filesObject on the DatatokenCreateParams

Comment (paulo-ocean) [reply src/utils/Assets.ts:199]: > for question, I have created #1859 in case.

i see, but you're again doing a fixed match:  confidentialEVM === Shappire 
afaik, that might not always be the case

Comment (mariacarmina) [reply src/utils/Assets.ts:199]: I agree @paulo-ocean to have a stronger condition and include chain id as well. I wanted to raise an awareness that when creating a datatoken (either bundled or standalone) to encrypt the tx since we pass the files object, even though files object is already encoded in utf8 bytes, that was related to my question. I agree to also merge your PR and do this `sapphire.wrap()` thing separate in my PR if needed when creating a datatoken.

Comment (mariacarmina) [reply src/utils/Assets.ts:199]: updated my PR with the stronger condition and tests as well.

**Review by mariacarmina — APPROVED** [review]
I can approve, the only pending point that we discovered on this PR is tackled on https://github.com/oceanprotocol/ocean.js/pull/1859

**Review by jamiehewitt15 — APPROVED** [review]
Looks good to me

**Comment by paulo-ocean** [comment]
merging, any open point we can address on a separate PR, as we need to update cli and the flow was already tested anyway

**Pull request merged**
